libpromises/evalfunction.c

```diff
@@ -923,6 +923,162 @@ static FnCallResult FnCallGetUsers(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const
 
 /*********************************************************************/
 
+
+#if defined(HAVE_GETPWENT) && !defined(__ANDROID__)
+
+static FnCallResult FnCallFindLocalUsers(EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
+{
+    assert(fp != NULL);
+    bool allocated = false;
+    JsonElement *json = VarNameOrInlineToJson(ctx, fp, finalargs, false, &allocated);
+    
+    // we failed to produce a valid JsonElement, so give up
+    if (json == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Function '%s' couldn't parse argument '%s'",
+            fp->name, RlistScalarValueSafe(finalargs));
+        return FnFailure();
+    }
+    else if (JsonGetElementType(json) != JSON_ELEMENT_TYPE_CONTAINER)
+    {
+        Log(LOG_LEVEL_ERR, "Bad argument '%s' in function '%s': Expected data container or slist",
+            RlistScalarValueSafe(finalargs), fp->name);
+        JsonDestroyMaybe(json, allocated);
+        return FnFailure();
+    }
+   
+    JsonElement *parent = JsonObjectCreate(10);
+    setpwent();
+    struct passwd *pw;
+    while ((pw = getpwent()) != NULL)
+    {
+        JsonIterator iter = JsonIteratorInit(json);
+        bool canAddToJSON = true;
+        JsonElement *element = JsonIteratorNextValue(&iter);
+        while (element != NULL)
+        {
+            if (JsonGetElementType(element) != JSON_ELEMENT_TYPE_PRIMITIVE) {
+                Log(LOG_LEVEL_ERR, "Bad argument '%s' in function '%s': Filter cannot include nested data",
+                    RlistScalarValueSafe(finalargs),fp->name);
+                return FnFailure();
+            }
+            const char *field = JsonPrimitiveGetAsString(element);
+            const Rlist *tuple = RlistFromSplitString(field, '=');
+            assert(tuple != NULL);
+            const char *attribute = TrimWhitespace(RlistScalarValue(tuple));
+
+            if (tuple->next == NULL)
+            {
+                Log(LOG_LEVEL_ERR, "Invalid filter field '%s' in function '%s': Expected attributes and values to be separated with '='",
+                    fp->name, field);
+                return FnFailure();
+            }
+            const char *value = TrimWhitespace(RlistScalarValue(tuple->next));
+
+            if (StringEqual(attribute, "name"))
+            {
+                if (!StringMatchFull(value, pw->pw_name))
+                {
+                    canAddToJSON = false;
+                }
+            }
+            else if (StringEqual(attribute, "uid"))
+            {
+                char uid_string[PRINTSIZE(pw->pw_uid) + 1];
+                int val = snprintf(uid_string, sizeof(uid_string), "%u", pw->pw_uid);
+                
+                if (val < 0)
+                {
+                    Log(LOG_LEVEL_ERR, "Couldn't convert the uid of '%s' to string in function '%s'",
+                        pw->pw_name, fp->name);
+                    return FnFailure();
+                }
```

Comment:
I would still do this as a sanity check
```suggestion
                }
                assert((size_t) val < sizeof(uid_string));
```